### PR TITLE
Implement get_winsize() for unix.

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by amaranth** [comment]
I think it'd be nicer if you had a const TIOCGWINSZ for the magic value. Also, is there a chance the ioctl could fail but write garbage in to the struct? That would currently get past your test but the call to ioctl returns -1 on error which should be a clearer indicator of a problem.

**Comment by dgriffen** [comment]
I missed the error return, that would be better. I'm not sure defining the const is necessary for this one use.

**Comment by nikomatsakis** [comment]
Is there any way to write a test for this?

**Comment by dgriffen** [comment]
I can't really think of a way to write a test for this since most tests will not take place in a tty. Not to mention if the test did take place in a tty it would have to always be of the same size for the test to work correctly.

**Comment by tbu-** [comment]
@dgriffen Defining the constant gives later readers something to google for.

File: src/libstd/sys/unix/tty.rs [hunk 1]
```diff
@@ -42,7 +42,26 @@ impl TTY {
...
+            #[repr(C)]
+            struct winsize {
+                ws_row: u16,
+                ws_col: u16,
+                ws_xpixel: u16,
+                ws_ypixel: u16
+            }
+
+            let mut size = winsize { ws_row: 0, ws_col: 0, ws_pixel: 0, ws_ypixel: 0 };
+            if funcs::bsd44::ioctl(0, 0x00005413, &mut size) == -1 {
```
Comment (alexcrichton) [inline src/libstd/sys/unix/tty.rs]: Can you move this literal into a `const` with a corresponding name that it has in C?

**Comment by alexcrichton** [comment]
Thanks @dgriffen! Just one minor nit (that was also previously mentioned!) and otherwise r=me

**Comment by utkarshkukreti** [comment]
Is this supposed to work on OSX? The value of `TIOCGWINSZ` on OSX is not the same as Linux.

File: src/libstd/sys/unix/tty.rs [hunk 2]
```diff
@@ -42,7 +44,26 @@ impl TTY {
...
+            #[repr(C)]
+            struct winsize {
+                ws_row: u16,
+                ws_col: u16,
+                ws_xpixel: u16,
+                ws_ypixel: u16
+            }
+
+            let mut size = winsize { ws_row: 0, ws_col: 0, ws_pixel: 0, ws_ypixel: 0 };
+            if funcs::bsd44::ioctl(0, TIOCGWINSZ, &mut size) == -1 {
```
Comment (utkarshkukreti) [inline src/libstd/sys/unix/tty.rs]: `s/0/self.fd.fd()`?

**Comment by alexcrichton** [comment]
ping @dgriffen, any thoughts on @utkarshkukreti's comments?

**Comment by dgriffen** [comment]
I don't have a mac to test, so i can't confirm this. But it appears TIOCGWINSZ only has a different value on powerpc macs. I could include the other value with a compiler directive. But i would like someone to confirm the value first.

**Comment by alexcrichton** [comment]
The value I have locally (OSX 10.10) for TIOCGWINSZ is `0x40087468`, and I also just checked that a Ubuntu server has the value `0x00005413` (as listed)

**Comment by dgriffen** [comment]
Is you mac a powerpc or intel?

**Comment by alexcrichton** [comment]
I'm using an intel mac

**Comment by dgriffen** [comment]
Updated the code with macos value. I cannot confirm that it works on mac, for aforementioned reasons. I have screenshots of a test on arch linux running gnome terminal. 


test.rs code is [here](http://pastebin.com/63vJ0tAX)

**Comment by alexcrichton** [comment]
Thanks! Could you squash the commits together as well?

**Comment by dgriffen** [comment]
Squashed everything together, should be ready to go.

File: src/libstd/sys/unix/tty.rs [hunk 3]
```diff
@@ -19,6 +19,12 @@ pub struct TTY {
     pub fd: FileDesc,
 }
 
+#[cfg(target_os = "macos")]
+const TIOCGWINSZ: i32 = 0x40087468;
```
Comment (utkarshkukreti) [inline src/libstd/sys/unix/tty.rs]: I think this (and the one below) should be of type `c_ulong`, not `i32`. Typo?

**Comment by dgriffen** [comment]
the ioctl function expects an i32, c_ulong is 64bit on 64bit machines, and has no way of being passes into ioctl.

**Comment by utkarshkukreti** [comment]
Apparently `liblibc` defines `ioctl` to take [`c_ulong` on Mac/BSDs](https://github.com/rust-lang/rust/blob/6869645e86c91544b8737b89809bdf10bef536d9/src/liblibc/lib.rs#L4809) and [`c_int` on Linux/Android](https://github.com/rust-lang/rust/blob/6869645e86c91544b8737b89809bdf10bef536d9/src/liblibc/lib.rs#L4843).

**Comment by dgriffen** [comment]
Also, it looks like openbsd used 0x40087468 as the value for TIOCGWINSZ, so I'll update that as well.

**Comment by alexcrichton** [comment]
Thanks!

**Comment by dgriffen** [comment]
I'll add android to the list for TIOCGWINSZ. Are there any other targets I need to add?

**Comment by dgriffen** [comment]
If I am unable to find a value for a specific os, should i leave code for that os as unimplemented?

**Comment by alexcrichton** [comment]
I found this locally:

```
./ndk_standalone/sysroot/usr/include/asm/ioctls.h:#define TIOCGWINSZ 0x5413
```

Other platforms which are built include iOS and dragonfly. Perhaps this function could only be defined for supported platforms for now?

**Comment by dgriffen** [comment]
Made changes to the code that should allow it to compile on all platforms. For now ios and dragonfly remain unimplemented.

**Pull request merged**